src/mpi_t/mpit_impl.c

```diff
@@ -0,0 +1,206 @@
+{
+    int mpi_errno = MPI_SUCCESS;
+    cat_table_entry_t *cat;
+    int i, num_pvars, count;
+
+    cat = (cat_table_entry_t *) utarray_eltptr(cat_table, cat_index);
+    num_pvars = utarray_len(cat->pvar_indices);
+    count = len < num_pvars ? len : num_pvars;
+
+    for (i = 0; i < count; i++) {
+        indices[i] = *(int *) utarray_eltptr(cat->pvar_indices, i);
+    }
+
+    return mpi_errno;
+}
+
+int MPIR_T_cvar_handle_alloc_impl(int cvar_index, void *obj_handle, MPI_T_cvar_handle * handle,
+                                  int *count)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_T_cvar_handle_t *hnd;
+
+    MPIR_CHKPMEM_DECL(1);
+
+    cvar_table_entry_t *cvar = (cvar_table_entry_t *) utarray_eltptr(cvar_table, cvar_index);
+
+    /* Allocate handle memory */
+    MPIR_CHKPMEM_MALLOC(hnd, MPIR_T_cvar_handle_t *, sizeof(*hnd), mpi_errno,
+                        "control variable handle", MPL_MEM_MPIT);
+#ifdef HAVE_ERROR_CHECKING
+    hnd->kind = MPIR_T_CVAR_HANDLE;
+#endif
+
+    /* It is time to fix addr and count if they are unknown */
+    if (cvar->get_count != NULL)
+        cvar->get_count(obj_handle, count);
+    else
+        *count = cvar->count;
+
+    hnd->count = *count;
+
+    if (cvar->get_addr != NULL)
+        cvar->get_addr(obj_handle, &(hnd->addr));
+    else
+        hnd->addr = cvar->addr;
+
+    /* Cache other fields */
+    hnd->datatype = cvar->datatype;
+    hnd->scope = cvar->scope;
+
+    *handle = hnd;
+
+    MPIR_CHKPMEM_COMMIT();
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    MPIR_CHKPMEM_REAP();
+    goto fn_exit;
+}
+
+int MPIR_T_cvar_read_impl(MPI_T_cvar_handle handle, void *buf)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int i, count;
+    void *addr;
+    MPIR_T_cvar_handle_t *hnd = handle;
+
+    count = hnd->count;
+    addr = hnd->addr;
+    MPIR_Assert(addr != NULL);
+
+    switch (hnd->datatype) {
+        case MPI_INT:
+            for (i = 0; i < count; i++)
+                ((int *) buf)[i] = ((int *) addr)[i];
+            break;
+        case MPI_UNSIGNED:
+            for (i = 0; i < count; i++)
+                ((unsigned *) buf)[i] = ((unsigned *) addr)[i];
+            break;
+        case MPI_UNSIGNED_LONG:
+            for (i = 0; i < count; i++)
+                ((unsigned long *) buf)[i] = ((unsigned long *) addr)[i];
+            break;
+        case MPI_UNSIGNED_LONG_LONG:
+            for (i = 0; i < count; i++)
+                ((unsigned long long *) buf)[i] = ((unsigned long long *) addr)[i];
+            break;
+        case MPI_DOUBLE:
+            for (i = 0; i < count; i++)
+                ((double *) buf)[i] = ((double *) addr)[i];
+            break;
+        case MPI_CHAR:
+            MPL_strncpy(buf, addr, count);
+            break;
+        default:
+            /* FIXME the error handling code may not have been setup yet */
+            MPIR_ERR_SETANDJUMP1(mpi_errno, MPI_ERR_INTERN, "**intern", "**intern %s",
```

Comment:
MPIR_ERR code being added back.

Reply:
You mean we shouldn't use `MPI_ERR_INTERN`? Do you suggest we add `MPI_T_ERR_INTERN`?

Reply:
No, I am saying that you removed all `MPIR_ERR_*` macros in a previous commit, but then added some back in this one. https://github.com/pmodels/mpich/pull/4952/commits/668887caace536dcd368aa701a1f4b7fe447da12

Reply:
Fixed.